librocksdb_sys/crocksdb/c.cc

```diff
@@ -2915,4 +2923,250 @@ const char* crocksdb_pinnableslice_value(const crocksdb_pinnableslice_t* v,
   *vlen = v->rep.size();
   return v->rep.data();
 }
+
+struct crocksdb_user_collected_properties_t {
+  UserCollectedProperties* inner;
+  crocksdb_user_collected_properties_t(UserCollectedProperties* props) : inner(props) {}
+};
+
+void crocksdb_user_collected_properties_add(crocksdb_user_collected_properties_t* props,
+                                            const char* key, size_t key_len,
+                                            const char* value, size_t value_len) {
+  props->inner->emplace(std::make_pair(std::string(key, key_len), std::string(value, value_len)));
+}
+
+struct crocksdb_table_properties_collector_t : public TablePropertiesCollector {
+  crocksdb_table_properties_collector_context_t* rep_;
+
+  crocksdb_table_properties_collector_t(
+    crocksdb_table_properties_collector_context_t* context) : rep_(context) {}
+
+  virtual ~crocksdb_table_properties_collector_t() {
+    rep_->destructor(rep_);
+  }
+
+  virtual Status AddUserKey(const Slice& key,
+                            const Slice& value,
+                            EntryType entry_type,
+                            SequenceNumber seq,
+                            uint64_t file_size) override {
+    rep_->add_userkey(rep_,
+                      key.data(), key.size(),
+                      value.data(), value.size(),
+                      entry_type, seq, file_size);
+    return Status::OK();
+  }
+
+  virtual Status Finish(UserCollectedProperties* inner) override {
+    crocksdb_user_collected_properties_t props(inner);
+    rep_->finish(rep_, &props);
+    return Status::OK();
+  }
+
+  virtual UserCollectedProperties GetReadableProperties() const override {
+    UserCollectedProperties inner;
+    crocksdb_user_collected_properties_t props(&inner);
+    rep_->readable_properties(rep_, &props);
+    return inner;
+  }
+
+  const char* Name() const override {
+    return rep_->name(rep_);
+  }
+};
+
+struct crocksdb_table_properties_collector_factory_t : public TablePropertiesCollectorFactory {
+  crocksdb_table_properties_collector_factory_context_t* rep_;
+
+  crocksdb_table_properties_collector_factory_t(
+    crocksdb_table_properties_collector_factory_context_t* context) : rep_(context) {}
+
+  virtual ~crocksdb_table_properties_collector_factory_t() {
+    rep_->destructor(rep_);
+  }
+
+  virtual TablePropertiesCollector* CreateTablePropertiesCollector (
+    TablePropertiesCollectorFactory::Context ctx) override {
+    auto context = rep_->create_table_properties_collector(rep_, ctx.column_family_id);
```

Comment:
How about rename `crocksdb_table_properties_collector_context_t` to `crocksdb_table_properties_collector_impl_t`?